### PR TITLE
2.x: factor out inner classes from the base reactive types

### PR DESCRIPTION
**Pull request by akarnokd**
This PR factors out the anonymous inner classes from the base reactive types and introduces the appropriate classes and methods in `*Helper` enums.

### CODE REVIEW

File: src/main/java/io/reactivex/internal/functions/Functions.java [hunk 1]
```diff
@@ -240,4 +243,367 @@ public int compare(Object a, Object b) {
...
+            this.value = value;
+        }
+        
+        @Override
+        public boolean test(T t) throws Exception {
+            return Objects.equals(t, value);
+        }
+    }
+    
+    public static <T> Predicate<T> equalsWith(T value) {
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/functions/Functions.java:362]: Is there any reason to use these static factories? The nested classes could just be public (or promoted to top-level) and referenced directly.

**Comment by JakeWharton** [comment]
:+1:

File: src/main/java/io/reactivex/internal/functions/Functions.java (hunk 1 above)
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/functions/Functions.java:362]: To avoid public namespace cluttering.

Comment (akarnokd) [inline src/main/java/io/reactivex/internal/functions/Functions.java:362]: Please don't refactor this yet; working on repeatWhen/retryWhen fixes.

**Pull request merged**